### ISSUE TITLE
does xgcm handle models with a tri-polar?

### ISSUE THREAD

**Issue by navidcy** [xgcm/xgcm#194]

**Comment by jbusecke** [comment]
Hi @navidcy, thanks for using xgcm. Short answer: Yes. I have worked with several models with xgcm sucessfully. 

Do you have a concrete problem or example that you are concerned about?

**Comment by dcherian** [comment]
@jbusecke this came up in our Ocean section meeting call today. It'd be nice to see an example notebook analyzing output on a tripolar grid (I have never worked with these so I'm pretty confused on how to represent things in xgcm).

**Comment by jbusecke** [comment]
What does pop use? 

Happy to make a n example with MOM5/6, but I just have to say that I am quite backed up with stuff rn.

In principal it doesn’t matter what kind of grid you have if it fits into one “logical” tile. The only time when things get more complicated is smth like a cubed sphere grid I think. But Ryan has that implemented in xgcm too.

**Comment by raphaeldussin** [comment]
is the question related to the folding of the north boundary? Which would basically be a face connection of a face onto itself with direction change from one side of the pivot point to the other.
I don't think I've seen this option in the docs/code.

**Comment by dcherian** [comment]
@raphaeldussin I came to that conclusion too. And yes I hadn't seen that option and so wanted to see what @jbusecke was doing

**Comment by jbusecke** [comment]
Hmm maybe I have been thinking about this a bit to simplified (forgive me I stay as far away from the Arctic as possible hahaha).
So you are asking what happens when you want to do e.g. a y derivative across the 'seam'? That sounds like something that could be addressed with some of the cubed-sphere logic that @rabernat has built. I however know next to nothing about that part of the code, so perhaps this question is best deferred to Ryan?

**Comment by raphaeldussin** [comment]
that's exactly right, velocities reverse directions between the 2 parts of the fold. You need the information on the pivot point (index and point type). The most natural choice is to use a corner (f, q, psi,... depending on your convention) but there's configurations for which the pivot is on a V-point (ORCA05 if my memory is right). I think the face_connection logic can work if we build a virtual new face containing the "mirror" image of the northernmost row.

what do you think?

**Comment by rabernat** [comment]
> I think the face_connection logic can work if we build a virtual new face containing the "mirror" image of the northernmost row.

These kinds of things are already done for the LLC grids, so should definitely be possible!

**Comment by dcherian** [comment]
In my limited understanding, the `face_connections` logic will work. However, the output doesn't have an "faces" so to speak. So it will have to be a virtual face that xgcm keeps track of internally. 

Maybe xgcm needs a `Seam` concept with kwargs `row, pivot_index, pivot_point_type`.

**Comment by raphaeldussin** [comment]
I can give it a go, but cannot really commit to a timeline. One approach would be to create a function, say:

```python

def create_north_fold(grid, pivot_index, pivot_point='corner')

```

that returns a mirrored face and adds a face dimension. However since faces cannot have different horizontal dimensions, we would need to add a mirror image of the whole domain, right?
Which means that we'd carry the extra dim and potentially double the computation if we don't pay attention to restrict to one face.

I'm wondering if a better solution would be to create just the mirror row and concatenate (since dask can do it lazily, I can tell from my ASTE experience :) )

**Comment by dcherian** [comment]
Concatenation would be the easiest!

**Comment by Thomas-Moore-Creative** [comment]
Just a note (for myself and others) that in #257 we're trying to document how to apply xgcm to an ACCESS model grid which is MOM5 based on the [Murray (1996)](https://doi.org/10.1006/jcph.1996.0136) tripolar grid that has a bipolar Arctic region.

**Comment by dcherian** [comment]
I *think* we could handle it as a fancy boundary condition here (as @raphaeldussin pointed out, create a mirrored row and sample from that): https://github.com/xgcm/xgcm/blob/f362bf0a629712ac3f448fffb7f9be1bdc1c59bc/xgcm/grid.py#L329-L338

We could also extend `boundary` to accept `"tripole"` in addition to `"periodic"`

**Comment by jbusecke** [comment]
Wouldnt we have to invert velocities as well? Is #117 relevant here?

**Comment by Thomas-Moore-Creative** [comment]
I'm not sure how many tri-polar grids are [Murray (1996)](https://doi.org/10.1006/jcph.1996.0136) but here's his Figure 7 that is referenced in the [MOM5 manual ](https://mom-ocean.github.io/assets/pdfs/MOM5_manual.pdf)along with that manual's figure 1.1.

<img width="976" alt="Screen Shot 2020-10-10 at 8 11 59 am" src="https://user-images.githubusercontent.com/36463906/95632040-85406f80-0ad0-11eb-9c29-09b8ba20fb9a.png">
<img width="1369" alt="Screen Shot 2020-10-10 at 8 15 03 am" src="https://user-images.githubusercontent.com/36463906/95632152-bcaf1c00-0ad0-11eb-8cf4-5729e0d1a456.png">

**Comment by Thomas-Moore-Creative** [comment]
And Michael Winton goes into further detail of how Murray 1996 is applied in MOM at the bottom of the page here:
https://www.gfdl.noaa.gov/guide-to-grid-coupling-in-fms/

> When vector quantities are passed across this cut, the sign must be changed on both u and v components (mpp_domains also supports this). The processor domains must also have mirror partners and mpp_domains currently requires that there be a processor boundary between the last and first i-row elements along the cyclic boundary on the pseudo-east/west edge. Consequently, when using the tripolar grid, arrays of geographic fields must start at the longitude of one of the bipolar grid poles. We have chosen the Siberian pole for admittedly non-science reasons.

**Comment by dcherian** [comment]
GIven all the ongoing work, can tripolar grids  be supported though a `boundary` option?

**Comment by jbusecke** [comment]
Yes. After the boundary padding refactor that will just be a new padding function!

**Comment by jbusecke** [comment]
Ok, now that v0.7.0 is released, I am very curious to pick this back up. 

I think the actual code implementation will be quite easy now. I suggest the code lives in a `_pad_tripolar` function in `padding.py`. What we should discuss here is which option such a function needs to be sufficiently general for all existing tripolar setups. This will then determine how to best 'tell' xgcm that a tripolar grid is used.

So let me kick this off with a round of questions to you all: 

1. Are there different types of tripolar 'seams'? As far as I understand, the basic idea is to take the northmost row of data and flip it around the central point. Are there other funky setups we need to consider or is the folding point always given by Ny/2 (where Ny is the number of zonal grid points). 
3. How are velocities dealt with? Simply reverse sign of u/v component? We have some machinery for doing this kind of stuff already in the `_pad_faceconnections` function and could probably generalize that a bit.
4. Anything else that needs to be considered?

**Comment by briochemc** [comment]
Hi there! Great package! I'm yet to try to really use it but this (dealing with simple grid topologies such as folded around like for ACCESS) is the type of thing I'd love to forget worrying about.

I just wanted to chime in on the following question in case this has not been resolved and I can contribute something useful:

> 2. How are velocities dealt with? Simply reverse sign of u/v component (again from the answer above this seems to be true)? 

I think the solution is not always that trivial, depending on what you're doing, but perhaps you have already thought about this and you can safely ignore my ramblings. For example, if your 𝑢/𝑣 are defined on the cell faces, and you are trying to get the meridional gradients of 𝑢 and 𝑣 on the folded boundary, then you need to be careful about the indices you use for those −𝑢 and −𝑣. (By meridional, I mean in the 𝑣 direction of the grid). See image below, which shows the curvilinear grid on the left with some points and cells of interest, and the logical grid equivalent on the right. The point I'm trying to make is that even though cell 𝑖 faces cell 𝑗, the 𝑢 and 𝑣 of the "ghost" of cell 𝑗 (denoted with an upside-down 𝑗 on the logical grid on the right) are different from the −𝑢 and −𝑣 of the actual cell 𝑗.

![IMG_71DB51945E40-1](https://github.com/xgcm/xgcm/assets/4486578/85487016-ed6d-49f4-b380-6a6dfb070fd6)

A specific example might be clearer: Let's say you  want the meridional gradient of 𝑣 at point A (star marker), then you'd need −𝑣 from cell 𝑙 (and not cell 𝑗) in your finite difference scheme. Similarly, if you want the meridional gradient of 𝑢 at point B (cross marker), then you'd need to use −𝑢 from cell 𝑘 (and not cell 𝑗).

Not sure if this makes sense. I may also be completely misunderstanding how this all works, so apologies for the noise if that's the case.

**Comment by hdrake** [comment]
> For example, if your 𝑢/𝑣 are defined on the cell faces, and you are trying to get the meridional gradients of 𝑢 and 𝑣 on the folded boundary, then you need to be careful about the indices you use for those −𝑢 and −𝑣. (By meridional, I mean in the 𝑣 direction of the grid).

Isn't the entire point of the tripolar grid to move the poles onto land so you don't need to deal with the discontinuities there, because they are guaranteed to be masked anyways?

**Comment by briochemc** [comment]
> > Hi there! Great package! I'm yet to try to really use it but this (dealing with simple grid topologies such as folded around like for ACCESS) is the type of thing I'd love to forget worrying about.
> 
> Isn't the entire point of the tripolar grid to move the poles onto land so you don't need to deal with the discontinuities there?

Yes, that's the entire point, but you have to be careful when interpolating. Particularly at the seam, where interpolation requires filling the ghost/halo cells. This is not trivial as when you rotate the grid around one of the north poles, which acts as a pivot point, you need to accurately reverse and shift `i` and `j` indices, and flip the sign of some quantities like horizontal velocities for example. And these reversing and shifting depend on many things: the type of staggered grid (A, B, C, etc.), the face/center location convention (faces on the north/east like MOM, or south/west like Oceananigans for example), the location of the north pole (on cell center like for some NEMO grids I think), x-face (also called U-pivot IIRC, also the default Oceananigans Tripolar configuration), y-face, or corner (F-pivot, MOM and ACCESS-OM2, recently added to Oceananigans). But in short all I meant to say is there's more to "just flip `u` and `v`".

**Comment by hdrake** [comment]
I made some good progress – largely thanks to @claude – on the tripolar fold problem. See this PR: https://github.com/xgcm/xgcm/pull/711

The PR includes a new notebook in xgcm-examples, which aims to showcase how the fold works in several different ocean models that differ in their implementations of the fold logic: https://github.com/xgcm/xgcm-examples/blob/448eee28b4189d6ce957384deb74d257877b5f8a/06_tripolar_fold.ipynb

I was able to test on some publicly available MOM6 and NEMO solutions, but I don't have any Oceananigans.jl solutions handy to compare against. (Claude attempted to configure and run a [minimal working example](https://github.com/xgcm/xgcm-examples/compare/a7b50a7ce1079c8d409e45b5ca6ea00dbf88636c...448eee28b4189d6ce957384deb74d257877b5f8a#diff-3086e197ac3f0cc561273a2e2a342016e43c9abe3482a7a1245bdd8a260c89bc), but it would be nice to have a realistic configuration at ~0.25º resolution that can be directly compared to the other solutions.)

The preliminary results below seem to suggest the current implementation in the PR works for the NEMO and MOM6 and grid (you may have to zoom in to see the fold in the fine-resolution MOM6 grid), but not for the Oceananigans.jl one. (It is possible this is because the Oceananigans.jl code itself is bugged, I guess: https://github.com/CliMA/Oceananigans.jl/issues/4466).

<img width="1375" height="413" alt="Image" src="https://github.com/user-attachments/assets/18e654b4-375c-4d8b-b1ce-4199de81965d" />

<img width="1363" height="399" alt="Image" src="https://github.com/user-attachments/assets/f178c174-71e6-43cf-b37e-93d7085bfc91" />

<img width="1361" height="399" alt="Image" src="https://github.com/user-attachments/assets/da7f9fa0-0f2b-40e5-8a57-47068b69f4e8" />

Any thoughts, @simone-silvestri @glwagner @briochemc @NoraLoose ? Is there an up-to-date realistic Oceananigans.jl config you can point me towards – either as output or config script?

**Comment by simone-silvestri** [comment]
Hi @hdrake. I think we have solved all the issues related with the tripolar grid (apologies to have left that issue open 😅 ).
If you want a configuration that runs on a tripolar grid you can use this one degree configuration https://numericalearth.github.io/NumericalEarthDocumentation/stable/literated/one_degree_simulation/ and maybe increase a bit the resolution to a quarter degree if you want to have more details in tracer and velocity fields.

I do not have at hand high resolution tripolar fields, but I can give you the [output from a 1 degree OMIP](https://www.dropbox.com/scl/fi/4ijg2qca2vhvp09vk5x6s/orca_corrected_snow_kskew800_ksymm800_bih50days_checkpoint_iteration95667.jld2?rlkey=bhiw7whrbo2mg2p82jjgjo70y&st=3d9dxkv5&dl=0) run on the same NEMO ORCA1 grid. It is a bit large given it is the checkpoint of this simulation object. You should be able to navigate to the ocean with 
`file["simulation/model/ocean/model"].

**Comment by hdrake** [comment]
@simone-silvestri , awesome thanks! I guess I should wait for https://github.com/CliMA/Oceananigans.jl/pull/5099 to be actually merged before generated this example data?

Is the grid for the 1º OMIP run the same default as in the config you linked? The purpose of this comparison is to compare as many different implementations of the fold as possible, so I actually prefer an Oceananigans.jl solution with a different treatment than the ORCA1 grid!

**Comment by simone-silvestri** [comment]
Oh I haven't realized it was not merged yet! Anyways, that PR just fixes global reductions over a tripolar grid so it should not play a role here. 

> Is the grid for the 1º OMIP run the same default as in the config you linked?

No, it's actually a different grid. If you prefer a different fold type, the default config of NumericalEarth should be good since it pivots on U-points rather than on F-points like ORCA.

**Comment by hdrake** [comment]
My PR that extends support to tripolar grids is now ready for review: https://github.com/xgcm/xgcm/pull/711

I would really appreciate feedback on the implementation and anyone who wants to test it on their own model output!
